src/frontend/src/optimizer/rule/table_function_to_internal_backfill_progress.rs

```diff
@@ -155,30 +177,91 @@ impl TableFunctionToInternalBackfillProgressRule {
     }
 }
 
-struct BackfillInfo {
+enum BackfillInfo {
+    SnapshotBackfill(SnapshotBackfillInfo),
+    NoShuffleOrArrangementBackfill(NoShuffleOrArrangementBackfillInfo),
+}
+
+struct NoShuffleOrArrangementBackfillInfo {
     job_id: u32,
+    fragment_id: u32,
+    table_id: u32,
     row_count_column_index: usize,
+}
+
+struct SnapshotBackfillInfo {
+    job_id: u32,
     fragment_id: u32,
     table_id: u32,
+    row_count_column_index: usize,
+    epoch_column_index: usize,
```

Comment:
Seems the two enum variants are constructed and matched only in this file. In other words, outside this file, two variants are different in whether returning a some in method `epoch_column_index`.

If so, it's more straight forward to keep the original `BackfillInfo` and add an optional field `epoch_column_index: Option<usize>`.